index.ts

```diff
@@ -76,23 +99,97 @@ export const sandbox = {
-            StreamSpecification: {
-              StreamEnabled: true,
-              StreamViewType: 'NEW_AND_OLD_IMAGES',
-            },
-          })
+    const tableStreams: TableStreamItem[] = inv['tables-streams']
+    abortController = new AbortController()
+    if (tableStreams?.length) {
+      continueProcessingLoop = true
+      const ddbStreamsClient = new DynamoDBStreamsClient({
+        region: inv.aws.region,
+        endpoint: local.url,
+        credentials,
+      })
+      // Init table streams for those defined
+      await Promise.all(
+        tableStreams
+          .map(({ table }) => table)
+          .filter(
+            (item: string, index: number, tableNames: string[]) =>
+              tableNames.indexOf(item) === index
+          )
+          .map((table: string) =>
+            dynamodbClient.send(
+              new UpdateTableCommand({
+                TableName: client.name(table),
+                StreamSpecification: {
+                  StreamEnabled: true,
+                  StreamViewType: 'NEW_AND_OLD_IMAGES',
+                },
+              })
+            )
+          )
+      )
+      // Reset Stream defaults
+      for (const arcStream of tableStreams) {
+        shardMap[arcStream.table] = []
+        await resetTableStreams(
+          dynamodbClient,
+          ddbStreamsClient,
+          arcStream.table
         )
+      }
+
+      dynamoDbStreamsLoop = periodically(
+        () => streamLoop(ddbStreamsClient),
+        2000,
+        abortController.signal
       )
-    )
+    }
+
+    async function streamLoop(ddbStreamsClient: DynamoDBStreamsClient) {
+      if (!continueProcessingLoop) return
+      for (const key of Object.keys(shardMap)) {
+        if (shardMap[key].length) {
+          const shardItem = shardMap[key].pop()
+          if (!shardItem) continue
+          try {
+            const event = await ddbStreamsClient.send(
+              new GetRecordsCommand({
+                ShardIterator: shardItem.ShardIterator,
+              })
+            )
+            if (event.Records?.length) {
+              tableStreams
+                .filter((x) => x.table === key)
+                .map((x) => {
+                  invoke({
+                    pragma: 'tables-streams',
+                    name: x.name,
+                    payload: event,
+                  })
+                })
+            }
+
+            if (event.NextShardIterator) {
+              shardMap[key].push({
+                ShardIterator: event.NextShardIterator,
+              })
+            }
+          } catch (error) {
+            if (error instanceof TrimmedDataAccessException) {
+              console.log(error.name)
+            }
+
+            await resetTableStreams(dynamodbClient, ddbStreamsClient, key)
+          }
+        }
+      }
+      setTimeout(() => streamLoop(ddbStreamsClient), 2000)
+    }
   },
   async end() {
+    continueProcessingLoop = false
```

Comment:
```suggestion
```

Reply:
These are still needed. 

When the reset is triggered, there may still be queued instances of the loop function. They attempt to read the streams, but the container has already been killed or removed, which causes the aws functions to fail and brick the whole system

This just gives those last instances an safe exit. I tried using a try/catch previously, but they wouldn't catch the error.

Reply:
> When the reset is triggered, there may still be queued instances of the loop function.

That's because of the call to `setTimeout` that is at the end of the loop function, which you can remove now because it is handled by `periodically`.